File: android/src/main/kotlin/com/syncrhrohealth/bandpass_ble_connection/ble/BlePeripheral.kt
package com.syncrhrohealth.bandpass_ble_connection.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothDevice
import android.content.Context

// TODO: update to real sensor
class BlePeripheral(
    private val context: Context,
    private val device: BluetoothDevice,
    private val callBack: BleConnectionCallBack
) {
    private val connection = BleConnection(context, device, callBack)
    var batteryLevel: Int = 0
    var isCharging: Boolean = false

    var hwVersion: String = ""
    var fwVersion: String = ""
    var model: String = ""
    var serialNumber: String = ""

    fun requestToConnect(autoConnect: Boolean = false) {
        connection.requestToConnect(autoConnect)
    }

    fun reconnect() {
        // reconnect device
        connection.reconnect()
    }

    fun requestToDisconnect() {
        connection.requestToDisconnect()
    }

    fun write(data: ByteArray) {
        connection.write(data)
    }

    fun write(data: String) {
        connection.write(data)
    }

    fun getAddress(): String {
        return device.address
    }

    fun readBatteryLevel() {
        connection.readBatteryLevel()
    }

    /**
     * bluetooth off may cause device name is null, cause NullPointerException
     */
    @SuppressLint("MissingPermission")
    fun getName(): String {
        return device.name;
    }
}

File: ios/Classes/Ble/BlePeripheral.swift
//
//  BlePeripheral.swift
//  Runner
//
//  Created by Vo Khac Tuyen on 9/3/25.
//

import Foundation
import CoreBluetooth


class BlePeripheral {
    private let connection: BleConnection
    private var device: CBPeripheral
    
    var batteryLevel: Int = 0
    var isCharging: Bool = false

    var hwVersion: String = ""
    var fwVersion: String = ""
    var model: String = ""
    var serialNumber: String = ""

    init(peripheral: CBPeripheral, callback: BleConnectionToDeviceCallback, central: BleCentralManager) {
        connection = BleConnection(peripheral: peripheral, callback: callback, central: central)
        device = peripheral
    }

    func connect() {
        connection.connect()
    }
    
    func getConnection() -> BleConnection {
        return connection
    }

    func requestToDisconnect() {
        connection.requestToDisconnect()
    }
    
    func changeDisconnectStatus() {
        connection.changeDisconnectStatus()
    }

    func write(data: Data) {
        connection.write(data: data)
    }

    func write(data: String) {
        
    }

    func getAddress() -> String {
        return device.identifier.uuidString;
    }

    func getName() -> String {
        return device.name ?? "";
    }
    
    func getDevice() -> CBPeripheral {
        return device
    }
}
